### ISSUE TITLE
cannot unmarshal DNS message

### ISSUE THREAD

**Issue by dinamic** [docker/for-mac#1103]
### Expected behavior

Bringing my project to life with docker-compose should work.

### Actual behavior

docker-compose fails.

### Information

Diagnostic ID: 008AC048-1D8B-426D-9D4E-AD570B959050

```
docker-compose up 
Creating network "sf2_default" with the default driver
Pulling data (busybox:latest)...
latest: Pulling from library/busybox
fdab12439263: Pulling fs layer
ERROR: error pulling image configuration: Get https://dseasb33srnrn.cloudfront.net/registry-v2/docker/registry/v2/blobs/sha256/1e/1efc1d465fd6255396d0efd90a899f57aba2b7b294b02d5f55c6f5505ca9f3e5/data?Expires=1483486613&Signature=QaCSfUKdsUkAt7ZtcVgcHNejBStOi5CytzSr~rZCmw9yiINywq9Oy4UX3-wrgiKm2f3IHEWTHXnl-1rdY~oLmTmqwvh-QFL9kHb1VhcJzgHJVsmiFuBzpimCtObY4MQkrWqBecFuu8xBbgQXue1M1diuBl2xnUicZczOUp7DFKY_&Key-Pair-Id=APKAJECH5M7VWIS5YZ6Q: dial tcp: lookup dseasb33srnrn.cloudfront.net on 192.168.65.9:53: cannot unmarshal DNS message
```

```
Docker for Mac: version: 1.12.5 (3e6f00c1d)
OS X: version 10.12.2 (build: 16C67)
logs: /tmp/008AC048-1D8B-426D-9D4E-AD570B959050/20170104-012112.tar.gz
[OK]     vmnetd
[OK]     dns
[OK]     driver.amd64-linux
[OK]     virtualization VT-X
[OK]     app
[OK]     moby
[OK]     system
[OK]     moby-syslog
[OK]     db
[OK]     env
[OK]     virtualization kern.hv_support
[OK]     slirp
[OK]     osxfs
[OK]     moby-console
[OK]     logs
[OK]     docker-cli
[OK]     menubar
[OK]     disk
```

```
# docker info
Containers: 2
 Running: 0
 Paused: 0
 Stopped: 2
Images: 13
Server Version: 1.12.5
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 32
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: host overlay bridge null
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 4.4.39-moby
Operating System: Alpine Linux v3.4
OSType: linux
Architecture: x86_64
CPUs: 4
Total Memory: 1.951 GiB
Name: moby
ID: YPMF:PT2C:UJZR:DRDD:A6OD:2NTF:UISY:KHBS:W6VL:JEBX:5ICR:5OWG
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 16
 Goroutines: 27
 System Time: 2017-01-03T23:25:34.705069763Z
 EventsListeners: 1
No Proxy: *.local, 169.254/16
Registry: https://index.docker.io/v1/
WARNING: No kernel memory limit support
Insecure Registries:
 127.0.0.0/8
```
```
# docker-compose -v
docker-compose version 1.9.0, build 2585387
```

Log files: 
- [docker.txt](https://github.com/docker/for-mac/files/683537/docker.txt)

### Steps to reproduce the behavior

1. create docker-compose.yml with the following content:

```
version: '2'

services:
    data:
        image: "busybox"
```

2. issue `docker-compose up`

Same problem occur trying to manually issue `docker pull busybox`.

```
# docker pull busybox

Using default tag: latest
latest: Pulling from library/busybox
fdab12439263: Pulling fs layer 
error pulling image configuration: Get https://dseasb33srnrn.cloudfront.net/registry-v2/docker/registry/v2/blobs/sha256/1e/1efc1d465fd6255396d0efd90a899f57aba2b7b294b02d5f55c6f5505ca9f3e5/data?Expires=1483487249&Signature=RXKxakGjZGYBnyBXxiXrr3SrtgeAUL6H6kbHYGEs1pCU8jDljfkzRJRp1d9tEKR5HlQigltqDz711QhEEusOm9eDbq9NlLBSoYufRSU7pnR41pDw7zPtmdCYslMu21D60K0kKlhvP6xI3Rx8ndl0muH79pDxhaY~7jyO2nJ2d9k_&Key-Pair-Id=APKAJECH5M7VWIS5YZ6Q: dial tcp: lookup dseasb33srnrn.cloudfront.net on 192.168.65.9:53: cannot unmarshal DNS message
```

**Comment by ebuildy** [comment]
Hello,

Don't have this problem on my D4M (Version 1.12.5 (14777)), ``docker pull busybox`` works fine.

Can you dig fine dseasb33srnrn.cloudfront.net on your machine and on D4M VM?

**Comment by dinamic** [comment]
So, the problem is present with the ISP at home, but not when I'm at work.

I will try `dig` when I'm back, but there should be no problem with that.

To me this seem more like reappearance of https://github.com/docker/docker/issues/24344.

**Comment by djs55** [comment]
There have been some DNS fixes in the beta branch which have not (yet) made their way to stable. If you have the time, could you try with the latest beta (34.1)? It's available from https://download.docker.com/mac/beta/Docker.dmg

**Comment by dinamic** [comment]
I will try it tonight, @djs55. Thanks!

**Comment by dinamic** [comment]
@djs55, it does not seem to work. :-/

The error message seem a bit different tho.

```
# docker pull busybox

Using default tag: latest
latest: Pulling from library/busybox
fdab12439263: Pulling fs layer 
error pulling image configuration: Get https://dseasb33srnrn.cloudfront.net/registry-v2/docker/registry/v2/blobs/sha256/1e/1efc1d465fd6255396d0efd90a899f57aba2b7b294b02d5f55c6f5505ca9f3e5/data?Expires=1483661206&Signature=ZCXq3SmHaPLXvMBCeYWjqhPsPKwaFEifcL629umEZ6oSUwM2UY00fr1lAxo3OP2-f8h9wQZe9O1OIXiIhSKzBsCcxx2mL14cLMGTyKk4BHKgGWpGdtfS0HwMbksffcFqVDI~gK1IcNHhY2t5~HfWRUYWn02xYhJbemgeD1iP6xk_&Key-Pair-Id=APKAJECH5M7VWIS5YZ6Q: dial tcp: lookup dseasb33srnrn.cloudfront.net on 192.168.65.1:53: server misbehaving
```

The hostname seems to be resolvable on the host OS:

```
# host dseasb33srnrn.cloudfront.net
dseasb33srnrn.cloudfront.net has address 54.230.202.50
dseasb33srnrn.cloudfront.net has address 54.230.202.174
dseasb33srnrn.cloudfront.net has address 54.230.202.171
dseasb33srnrn.cloudfront.net has address 54.230.202.205
dseasb33srnrn.cloudfront.net has address 54.230.202.120
dseasb33srnrn.cloudfront.net has address 54.230.202.110
dseasb33srnrn.cloudfront.net has address 54.230.202.142
dseasb33srnrn.cloudfront.net has address 54.230.202.7
```

Let me know what details would be useful to troubleshoot this one.

**Comment by dinamic** [comment]
@djs55, do you need more information on this?

**Comment by djs55** [comment]
Thanks for the reminder and the new diagnostics. Looking at the logs I think the error is now the same as https://github.com/docker/for-mac/issues/1025 i.e. we send the query to all your servers in parallel, and take the first response. Unfortunately the first response seems to be a "Refused". I've changed the policy in the master branch to wait for a successful response and only return errors like "Refused" if all the servers fail.

Unfortunately this fix is queued up behind the 1.13 stable release (which itself was queued behind the 1.12.6 security fix). Once 1.13 hits the stable channel (which is hopefully imminent now) this should be released in the next beta. Separately we're working to create a "nightly" channel which would be populated by builds from master which pass our regression tests. I'm sorry for the delay on this.

If you'd like to try something highly experimental then here is a replacement for the networking component, built from the master branch:

[slirp.zip](https://github.com/docker/for-mac/files/704055/slirp.zip)

```
$ sha1sum slirp.zip 
3ece6d4fa9c06fce3e5bbabe5538bd081580aa54  slirp.zip
```

To try it, execute the following: (assuming your `Docker.app` is in `/Applications`)
```
$ unzip slirp.zip 
Archive:  slirp.zip
  inflating: com.docker.slirp        
$ cp com.docker.slirp /Applications/Docker.app/Contents/MacOS/com.docker.slirp
```
and then restart the app.

Let me know if this makes a difference. If not, could you upload a fresh set of diagnostics and I'll have a look to see what went wrong. I've mostly tested this with component-level simulations so it's possible there's still a case I've missed. Any feedback would be very helpful. Separately we're developing an "enterprise" testing network for end-to-end testing where everything is configured to be maximally hostile e.g. no direct HTTP outgoing except through authenticated proxies etc etc.

Thanks again for your report!

**Comment by dinamic** [comment]
@djs55, no worries. I can always input Google's DNS servers and it would work just fine. Is just that this seem to be an isolated case that happens only in some specific conditions. As a software developer myself I know it may be hard to pinpoint the reason for a problem if you don't have enough data, so if I can help reproduce the problem and give you enough data - why not, right.

I mean, is not a blocker for me if this does not work as it has its workaround. Is just that it would be nice to have it resolved.

Thanks for sharing all the details you've shared!
I will try out the `slirp.zip` in the upcoming days and will let you know how it goes.

**Comment by dinamic** [comment]
@djs55 This fixes the issue for me as well.

I have sent a diagnostic nevertheless: `3A0E8B41-8055-4276-BC6D-C9D54730829E`

```
# docker pull busybox

Using default tag: latest
latest: Pulling from library/busybox
557a0c95bfcd: Pull complete
Digest: sha256:ae007bdb45fc0d56e3d705b97640ac24844bcc9ce4c8b8493f216a57ab6af0d5
Status: Downloaded newer image for busybox:latest
```

**Comment by KirillGrishin** [comment]
@djs55 Using that `slirp.zip` that you posted above seemed to have solved the problem (I am using Docker 1.13.0 on Mac).

Just for other googlers I will leave here the error that I was having:
`Error response from daemon: Get https://registry-1.docker.io/v2/: dial tcp: lookup registry-1.docker.io on 192.168.65.1:53: server misbehaving`

**Comment by djs55** [comment]
Thanks everyone for the feedback. I've tentatively marked this ticket as fixed (since the patch is in the master branch). I'll let you know when it ships in an official beta.

**Comment by mattdavenport** [comment]
Adding to this as the issue may be related, the fix from @djs55 worked great 👍. However I just tried to log in (via basic auth) to a custom image registry and the same problem arose. It appears that Docker (post-patch above) isn't using the same DNS resolution/method for login as it is for pulling images. Perhaps the same fix could be duplicated in the other command. Not familiar enough with the code base to elaborate further though..

**Comment by semifor** [comment]
This fix was working perfectly for me last Friday. Today it's not. I've discovered that it's working for the default docker0 bridge network, but not for the default bridge network set up by docker-compose for my application.

I expect both of these commands to produce a similar result. They don't.

```
➜  docker run --rm -ti app_sphinx cat /etc/resolv.conf
# Generated by dhcpcd from eth0.dhcp
# /etc/resolv.conf.head can replace this line
domain elided.private.datacenter.vpn
nameserver 192.168.65.1
# /etc/resolv.conf.tail can replace this line
➜  docker-compose run --rm app_sphinx cat /etc/resolv.conf
nameserver 127.0.0.11
options ndots:0
```

**Comment by semifor** [comment]
@ebuildy my attempt to describe the problem was bad. My apologies for that.

When the problem exists, I can ping a host across the vpn from my osx command line. I can ping it from a container running the default0 bridge network. I cannot ping it from a container on a default private network created via `docker-compose up`. Instead, I get "bad address".

When it's working normally, I'm able to resolve hosts across the vpn from a container in the default private network. I haven't found a reliable way to reproduce the problem. When the problem exists, eventually restarting docker engine, the vpn connection, and the services works. I don't yet know if it's order dependent, or if it simply started working after multiple tries.

**Comment by samoht** [comment]
I am going to close that issue as the original problem is now fixed.

@mattdavenport your problem seems to be similar to https://github.com/docker/for-mac/issues/1220. I am not familiar enough with `docker login` to help, but I will try to ping someone who knows better. Please comment on that #1220  if you have more details (or if you manage to understand the issue better).

@semifor would you mind opening a new issue with a diagnostic ID and reproduction steps so that we can track/follow-up with your current issue?  Thanks!

**Comment by KirillGrishin** [comment]
@samoht I am not sure if the fix is included in 1.13.1, but the issues is still there.

```
mes-MacBook-Pro:~ megrishin$ docker --version
Docker version 1.13.1, build 092cba3
mes-MacBook-Pro:~ megrishin$ docker pull redis
Using default tag: latest
Error response from daemon: Get https://registry-1.docker.io/v2/: dial tcp: lookup registry-1.docker.io on 192.168.65.1:53: server misbehaving
mes-MacBook-Pro:~ megrishin$
```

**Comment by mthole** [comment]
@djs55 Has this shipped in an official beta yet? It's not immediately clear to me how to find out myself.

**Comment by Novitoll** [comment]
+1 same issue in lInux with cisco anyconnect vpn
